### PR TITLE
Improve operator handling

### LINKED ISSUE #410: Add brackets when mixing `and` and `or` operations for stylistic reasons
for this python code : 
```
if (x >= 0 and x <= 5) or x == 10:
        x = x + 1
```

the translated code was like this : 

```
if (x >= 0 && x <= 5 || x == 10)
    {
        x = x + 1;
    }
```

even if there is no brackets this code is working well, but it raises a warning when compiling the generated code in C, so we must add brackets when there are mixed Or and And.

----------------------------------------------------------

> It is working as intended for both. The intention is to add brackets when they are necessary in python.
> As I said above, we have 3 cases here:
> 1. `(x >= 0 and x <= 5) or x == 10`
> 2. `x >= 0 and (x <= 5 or x == 10)`
> 3. `x >= 0 and x <= 5 or x == 10`
> 
> Case 2 and 3 will give the same results in python as `and` operations  are always carried out before `or` operations when there are no brackets.
> The parsing step does not read the brackets (and we can't make it read them) so we add them back in where they are necessary. In case 2 they are not necessary. While in case 1 they are.
> 
> That said option 2 is a lot easier to understand than option 3. This is why C raises a warning when you don't use brackets. As a result it may be a good idea to always use brackets when mixing `and` and `or` operations for stylistic reasons. But this should be discussed in an issue. The only people who see messages in pull requests are people that participated in the Pull Request and it would be good to have other opinions too. Please copy this conversation into an issue

_Originally posted by @EmilyBourne in
https://github.com/pyccel/pyccel/pull/396#issuecomment-681101777_

### LINKED ISSUE #529: Fix Fortran printer of numpy.random.randint
Bug description
-----------------
The Fortran printer generates invalid code for `numpy.random.randint` if the `low` parameter is an integer literal.

To reproduce
-----------------
The Python code
```python
import numpy as np

x = np.random.randint(-10, 10)
```
is translated to
```fortran
integer(kind=8) :: x
real(kind=8) :: Dummy_0001

call random_number(Dummy_0001)

x = floor((10_8 - -10_8) * Dummy_0001 + -10_8, kind=8)
```
which cannot compile because of two consecutive signs (`- -` and `+ -`).

Expected behavior
-----------------
The last line of code should be
```fortran
x = floor((10_8 - (-10_8)) * Dummy_0001 + (-10_8), kind=8)
```
or, much better,
```fortran
x = floor((10_8 + 10_8) * Dummy_0001 - 10_8, kind=8)
```

Language
-----------------
Fortran

Additional context
-----------------
This is possibly caused by an issue with the AST node `PyccelMinus`, or its printer function.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Create an `ast.operators` file containing all python operators. Operators in this file have a name beginning with `Pyccel`. Class inheritance is used abundantly to reduce code duplication. Precedence handling is now carried out at all stages in case instances are created on the fly (fixes problem reported in PR 410)

**Commit Summary**
- Move operators into an `ast.operators` file
- Rename `PyccelOperator`->`PyccelBinaryOperator`. Create super class `PyccelOperator`
- Create super classes `PyccelBooleanOperator`, `PyccelComparisonOperator`, `PyccelUnaryOperator`, `PyccelBitOperator`, `PyccelArithmeticOperator`
- Split `__init__` into 3 steps:
   1. `_handle_precedence`
   2. `_set_dtype`
   3. `_set_shape_rank`
- Define `_set_dtype` and `_set_shape_rank` in specialized classes
- Handle repeated unary operators (This stops ++a being written)
- Add specialised `_handle_precedence` method to `PyccelAnd`, `PyccelOr`, and `PyccelArithmeticOperator`. Fixes #410
- Rename to match naming convention `Is` -> `PyccelIs`, `IsNot` -> `PyccelIsNot`
- Simplify `sympy_to_pyccel` function using the fact that `PyccelAdd`/`PyccelMinus` is always binary¨
- Remove unnecessary `PyccelMinus` inheritance from `PyccelAdd` (`PyccelAdd` is more specialised than `PyccelMinus`)
- Use new superclasses to simplify semantic parser
- Add test to show PR fixes #529

### CODE REVIEW

**Review by saniQQ — APPROVED** [review]

**Review by nhamidn — APPROVED** [review]

File: pyccel/ast/operators.py [hunk 1]
```diff
@@ -0,0 +1,949 @@
+"""
+Module handling alll python builtin operators
```
Comment (yguclu) [inline pyccel/ast/operators.py:2]: Typo...

**Review by yguclu — APPROVED** [review]

File: pyccel/ast/operators.py [hunk 2]
```diff
@@ -0,0 +1,949 @@
...
+
+                self._shape = shape
+                self._rank  = len(shape)
+            else:
+                self._rank  = max(a.rank for a in self._args)
+                self._shape = [None]*self._rank
+
+#==============================================================================
+
+class PyccelArithmeticOperator(PyccelBinaryOperator):
```
Comment (EmilyBourne) [inline pyccel/ast/operators.py:391]: Is anything a `PyccelBinaryOperator` but not a `PyccelArithmeticOperator`?

Comment (EmilyBourne) [reply pyccel/ast/operators.py:391]: Yes. Comparisons

Comment (yguclu) [reply pyccel/ast/operators.py:391]: There is a `PyccelComparisonOperator`...

**Pull request merged**